# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/SharedFontList.cpp
# Commit: 3acfcd288f56
# Full Hash: 3acfcd288f567c09ce806ce144f48447bd01b661
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2019-08-30 21:54:33
# Description:
#   Bug 1576846 - Fix the content-process's use of mAliasTable to temporarily record alternate font family names on Windows. r=jwatt
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D43746
# ==============================================================================

diff -r 3e9155a5cd24 -r 3acfcd288f56 gfx/thebes/SharedFontList.cpp
--- a/gfx/thebes/SharedFontList.cpp	Fri Aug 30 12:52:42 2019 +0000
+++ b/gfx/thebes/SharedFontList.cpp	Fri Aug 30 13:14:26 2019 +0000
@@ -757,7 +757,16 @@
   // style name from the end of the requested family name.
   // After the deferred font loader has finished, this is no longer needed as
   // the "real" family names will have been found in AliasFamilies() above.
-  if (!header.mAliasCount && aName.Contains(' ')) {
+  if (aName.Contains(' ')) {
+    auto pfl = gfxPlatformFontList::PlatformFontList();
+    if (header.mAliasCount) {
+      // Aliases have been fully loaded by the parent process, so just discard
+      // any stray mAliasTable and mLocalNameTable entries from earlier calls
+      // to this code, and return.
+      pfl->mAliasTable.Clear();
+      pfl->mLocalNameTable.Clear();
+      return nullptr;
+    }
     const nsLiteralCString kStyleSuffixes[] = {
         nsLiteralCString(" book"),   nsLiteralCString(" medium"),
         nsLiteralCString(" normal"), nsLiteralCString(" regular"),
@@ -776,10 +785,14 @@
           // (either it's already in mAliasTable, or it gets added there when
           // we call ReadFaceNamesForFamily on this candidate).
           Family* candidateFamily = &families[match];
-          auto pfl = gfxPlatformFontList::PlatformFontList();
           if (pfl->mAliasTable.Lookup(aName)) {
             return candidateFamily;
           }
+          // Note that ReadFaceNamesForFamily may store entries in mAliasTable
+          // (and mLocalNameTable), but if this is happening in a content
+          // process (which is the common case) those entries will not be saved
+          // into the shared font list; they're just used here until the "real"
+          // alias list is ready, then discarded.
           pfl->ReadFaceNamesForFamily(candidateFamily, false);
           if (pfl->mAliasTable.Lookup(aName)) {
             return candidateFamily;